plugins/storage/volume/default/src/main/java/org/apache/cloudstack/storage/datastore/driver/CloudStackPrimaryDataStoreDriverImpl.java

```diff
@@ -379,11 +382,21 @@ public void takeSnapshot(SnapshotInfo snapshot, AsyncCompletionCallback<CreateCm
 
     @Override
     public void revertSnapshot(SnapshotInfo snapshot, SnapshotInfo snapshotOnPrimaryStore, AsyncCompletionCallback<CommandResult> callback) {
-        RevertSnapshotCommand cmd = new RevertSnapshotCommand((SnapshotObjectTO)snapshot.getTO(), (SnapshotObjectTO)snapshotOnPrimaryStore.getTO());
+        SnapshotObjectTO dataOnPrimaryStorage = null;
+        if (snapshotOnPrimaryStore != null) {
```

Comment:
same as above if snapshotOnPrimaryStore is null then dataOnPrimaryStorage is never set? (or is it that this will never be null?)

Reply:
It can be null. We are handling the same case where snapshotOnPrimary is null.
Following is the complete flow 
1. Look for snapshotOnPrimary entry exists
2. If exists, use that entry to find the appropriate driver to execute revertSnapshotCommand
3. If not found, use volume information to find the appropriate driver to execute revertSnapshotCommand
4. While executing the revertSnapshotCommand, if snapshotOnPrimary exists, use it to revert the volume
5. If snapshotOnPrimary does not exists, use the snapshot from secondary storage to revert the volume